File: study/2023/02-W1/youjin.dart
class Solution {
  int findCheapestPrice(int n, List<List<int>> flights, int src, int dst, int k) {
    final routes = _routes(n, flights);
    final mapped = List.generate(n, (index) => double.maxFinite.toInt());
    mapped[src] = 0;

    final queue = [
      [src, 0, 0]
    ];

    while (queue.isNotEmpty) {
      final shiftQueue = queue.removeAt(0);
      final int now = shiftQueue[0];
      final int totalPrice = shiftQueue[1];
      final int move = shiftQueue[2];

      routes[now].forEach(([int? next, int? price]) => {

        if(price ==null){ price = 0 },

            if (move <= k && totalPrice + price < mapped[next!]){
                mapped[next] = totalPrice + price,
                queue.add([next, totalPrice + price, move + 1]),
              }
          });
    }

    return mapped[dst] != double.maxFinite.toInt() ? mapped[dst] : -1;
  }

  _routes(int n, List<List<int>> flights) {
    final list = List.generate(n, (index) => []);

    for (var ele in flights) {
      list[ele[0]].add([ele[1], ele[2]]);
    }

    return list;
  }
}

File: study/2023/02-W1/youjin.js
const findCheapestPrice = (n, flights, src, dst, k) => {
  const routes = _routes(n, flights);

  const mapped = Array.from({ length:n }, () => Number.MAX_SAFE_INTEGER);
  
  mapped[src] = 0;
  const queue = [[src, 0, 0]];
    
  while (queue.length > 0) {
      const [now, totalPrice, move] = queue.shift();

    routes[now].forEach(([next, price]) => {
      if (move <= k && totalPrice + price < mapped[next]) {
        mapped[next] = totalPrice + price;
        queue.push([next, totalPrice + price, move + 1]);
      }
    })
  };
  
  return mapped[dst] !== Number.MAX_SAFE_INTEGER ? mapped[dst] : -1;
};

const _routes = (n, flights) => {
	const arr = Array.from({ length: n }, () => []);
  
	flights.forEach((flight) => {
		const [start, end, price] = flight;
		arr[start].push([end, price]);
	});
  
	return arr;
};


// Runtime 102 ms
// Beats 74.52%
// Memory 46 MB
// Beats 74.90%
